File: SPOJ/PT07Z.cpp
#include <iostream>
#include <deque>
#include <set>
#include <map>
#include <vector>

using namespace std;

map<int, vector<int> > data;
map<int, int> parent;
set<int> visited;
deque<int> q;
int furthest = 1;

void bfs(int start_node) {
  parent.clear();
  visited.clear();
  q.clear();
  q.push_back(start_node);
  while(!q.empty()) {
    int cur_node = q.front();
    q.pop_front();
    visited.insert(cur_node);
    furthest = cur_node;
    map<int, vector<int> >::iterator iter = data.find(cur_node);
    if(iter != data.end()) {
      for(vector<int>::iterator it = iter->second.begin(); it != iter->second.end(); it++) {
        if(visited.find(*it) != visited.end())
          continue;
        map<int, int>::iterator cur_par = parent.find(cur_node);
        if((cur_par == parent.end()) || (cur_par->second != *it)) {
          parent[*it] = cur_node;
        }   
        q.push_back(*it);
      }
    }
  }
}

int find_path_length(int node) {
  map<int, int>::iterator node_par = parent.find(node);
  if(node_par == parent.end())
    return 0;
  else
    return 1 + find_path_length(node_par->second);
}

int main() {
  int nodes;
  cin>>nodes;
  if(nodes <= 1) {
    cout<<0<<endl;
    return 0;
  }
  int u, v;
  for(int i=0; i<nodes-1; i++) {
    cin>>u>>v;
    map<int, vector<int> >::iterator iter = data.find(u);
    if(iter != data.end())
      iter->second.push_back(v);
    else {
      vector<int> v1;
      v1.push_back(v);
      data[u] = v1;
    }
    iter = data.find(v);
    if(iter != data.end())
      iter->second.push_back(u);
    else {
      vector<int> v2;
      v2.push_back(u);
      data[v] = v2;
    }
  }
  bfs(1);
  bfs(furthest);
  cout<<find_path_length(furthest)<<endl;
  return 0;
}

File: SPOJ/PT07Z.rb
require 'set'
require 'thread'

num = gets.to_i
if num <= 1
  puts 0
  abort
end
$data = Hash.new
$parent = Hash.new
$queue = Queue.new
$visited = Set.new

# Get the input
(num-1).times do
  node1, node2 = gets.split(" ").map{ |i| i.to_i }
  if $data[node1] == nil
    $data[node1] = [node2]
  else
    $data[node1] << node2
  end
  if $data[node2] == nil
    $data[node2] = [node1]
  else
    $data[node2] << node1
  end
end

# BFS
def bfs(start_node)
  $queue.clear
  $visited.clear
  $parent.clear
  $queue.push start_node
  until $queue.empty?
    cur_node = $queue.pop
    # next if $visited.include? cur_node
    $visited.add cur_node
    $furthest = cur_node
    unless $data[cur_node].nil?
      $data[cur_node].each do |node|
        next if $visited.include? node
        $parent[node] = cur_node unless $parent[cur_node] == node
        $queue.push node
      end
    end
  end
end

def find_path_length(node)
  if $parent[node] == nil
    return 0
  else
    return 1 + find_path_length($parent[node])
  end
end

# start from node number 1
bfs(1)
# puts "Furthest is #{$furthest}"
# puts "Furthest cur_node is #{$cur_node}"

# Do another BFS and find furthest from here
bfs($furthest)
# puts "Furthest is #{$furthest}"
# puts "Furthest cur_node is #{$cur_node}"

#puts $parent.inspect
#puts $visited.inspect
#puts $data.inspect
#puts "Longest path length is #{find_path_length($furthest)}"
puts find_path_length $furthest
